kep/624-disk-io-aware-scheduling/README.md

```diff
@@ -0,0 +1,207 @@
+
+### IO Calculation Model
+
+The IO Calculation Model is responsible for converting the disk’s available IO capacity and each new pod’s IO request to normalized values. Since the normalization function is disk type and vendor specific, this component is outside the scope of the scheduler plugin’s implementation.
+
+### Disk IO Scheduler Plugin
+
+We leverage the K8s scheduler framework to add the disk IO scheduler plugin.
+When the scheduler plugin starts, it subscribes to the IO Driver with a long streaming connection. The IO Driver updates the scheduler plugin about each node disk’s normalized available IO capacity and the scheduler plugin stores the info in the plugin’s local cache.
+
+The disk IO scheduler plugin consists of the following parts.
+
+#### Filter Plugin
+
+During the filter phase, the scheduler plugin sends the PodSpec with a disk IO BW request (as shown below) to the IO Driver and gets back a normalized disk IO BW needed by this POD. It then loops through each node in the candidate node list and checks this needed disk IO request of the POD against each node’s available disk IO capacity saved in the local cache to generate an updated candidate list.
+
+``` yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: ga_pod
+  annotations:
+     blockio.kubernetes.io/throughput: |
+{"rbps": "20M","wbps": "30M","blocksize": "4k"}
+spec:
+  containers:
+    - name: xxxServer
+      image: xxx
+       volumeMounts:  
+        - name: xxx-storage
+          mountPath: /data/xxx
+  volumes:
+  - name: xxx-storage
+     emptyDir: {}
+```
+
+#### Score Plugin
+
+During the score phase, the scheduler plugin gives a score to each node in the candidate list based on a scoring policy.
+```
+T = Node's available disk IO capacity
+R = Pod's needed disk IO bandwidth
+```
+For the Most Allocated policy:
+
+$$ score = { T - R \over T} $$
+For the Least Allocated policy:
+$$ score = {R \over T} $$
+
+#### Reserve Plugin
+
+During the reserve phase, the scheduler plugin updates the selected node’s available disk IO capacity by deducting the pod’s needed disk IO resource. In addition, it adds the pod to the ReservedPods list, tags the list with a Generation, and then notifies the IO Driver of the new ReservedPods list. The Generation, similar to the Kubernetes’s [ResourceVersion](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions), ensures that only the update on the latest generation ReservedPod list is saved to the scheduler plugin’s cache.
+
+Whenever there is any change to the IO metric collection context, the Generation increases by 1. If the IO Driver reports data based on an older generation than what is saved in the scheduler plugin’s cache, the update will be discarded.
+<p align="center"><img src="images/reserve.png" title="Reserve Phase" width="600" class="center"/></p>
+The API protocol between the disk IO scheduler plugin and the IO Driver is outlined below:
```

Comment:
Which factors made you decide for this form of communication instead of using a CRD or a new kubernetes datatype in general to enable communication between components? Could you perhaps elaborate about pros/cons of the different approaches?

Reply:
Which service (a new daemonset perhaps?) will provide the data?

Reply:
> Which service (a new daemonset perhaps?) will provide the data?

The implementation is vendor-specific. One possible reference implementation involves an aggregator running on the control plane as a service to consolidate data collected from all IO metrics collectors, while IO metrics collectors run on worker nodes as daemonsets.

Reply:
> Which factors made you decide for this form of communication instead of using a CRD or a new kubernetes datatype in general to enable communication between components? Could you perhaps elaborate about pros/cons of the different approaches?

In our scenario, the disk IO request of a pod specified in the pod spec must be normalized. To obtain the normalized IO bandwidth, the scheduler plugin needs to call the EstimateRequest API and obtain normalized IO BW from IO driver. Using a Custom Resource Definition (CRD) has the advantage of allowing reuse of Kubernetes’ existing security mechanisms. However, the disadvantage is that using the CRD mechanism to obtain the normalized IO bandwidth can increase scheduling latency.